crates/sui-core/src/consensus_adapter.rs

```diff
@@ -244,12 +266,146 @@ impl ConsensusListener {
+            .map(|_| ())
+    }
+
+    /// Wait for a transaction to be sequenced by consensus (or to timeout).
+    async fn waiter<T>(
+        receiver: oneshot::Receiver<SuiResult<SerializedTransactionInfoResponse>>,
+        retry_delay: Duration,
+        deliver: T,
+    ) -> (SuiResult<SerializedTransactionInfoResponse>, T) {
+        let outcome = match timeout(retry_delay, receiver).await {
+            Ok(reply) => reply.expect("Failed to read back from consensus listener"),
+            Err(e) => Err(SuiError::FailedToHearBackFromConsensus(e.to_string())),
+        };
+        (outcome, deliver)
+    }
 
-    /// Hash serialized consensus transactions. We do not need specific cryptographic properties except
-    /// only collision resistance.
-    pub fn hash(serialized: &SerializedConsensusTransaction) -> ConsensusTransactionDigest {
-        let mut hasher = DefaultHasher::new();
-        serialized.hash(&mut hasher);
-        hasher.finish()
+    /// Main loop receiving checkpoint fragments to reliably submit to consensus.
+    async fn run(&mut self) {
+        let mut waiting = FuturesUnordered::new();
+
+        // Continuously listen to checkpoint fragments and re-attempt sequencing if needed.
+        loop {
+            // Try to submit all pending checkpoint fragments to consensus.
+            while let Some(serialized) = self.buffer.pop_back() {
+                match self.submit(serialized.clone()).await {
+                    Ok(_) => {
+                        // Notify the consensus listener that we wish to be notified once our
+                        // consensus transaction is sequenced.
+                        let (sender, receiver) = oneshot::channel();
+                        let consensus_input =
+                            ConsensusListenerMessage::New(serialized.clone(), sender);
+                        self.tx_consensus_listener
+                            .send(consensus_input)
+                            .await
+                            .expect("Failed to notify consensus listener");
+
+                        // Add the receiver to the waiter. So we can retransmit if the
+                        // connection fails.
+                        let future = Self::waiter(receiver, self.retry_delay, serialized);
+                        waiting.push(future);
+                    }
+                    Err(_) => {
+                        self.buffer.push_back(serialized);
+                        break;
+                    }
+                }
+            }
+
+            // Process new events.
+            tokio::select! {
+                // Listen to new checkpoint fragments.
+                Some(transaction) = self.rx_checkpoint_consensus_adapter.recv() => {
+                    if self.buffer.len() >= self.max_pending_transactions {
+                        // Drop the earliest fragments. They are not needed for liveness.
+                        let _ = self.buffer.pop_back();
```

Comment:
@gdanezis Is this true?

Reply:
I would instead recommend we cleanup fragments that might be stuck for previous checkpoints. If a fragment is for the current checkpoint (or future), it is not clear it is not clear it is ok to just drop.

Reply:
fixed